crates/syntax/src/ast/make.rs

```diff
@@ -149,8 +153,65 @@ pub fn impl_(
     ast_from_text(&format!("impl{} {}{} {{}}", params, ty, ty_params))
 }
 
-pub fn impl_trait(trait_: ast::Path, ty: ast::Path) -> ast::Impl {
-    ast_from_text(&format!("impl {} for {} {{}}", trait_, ty))
+pub fn impl_trait(trait_: &ast::Path, adt: &ast::Adt, code: &str) -> ast::Impl {
```

Comment:
Ah I think I misunderstood you here a bit. The `make` module itself shouldn't do anything fancy aside from stupidly creating syntax trees.

So the signature of this function here should be something like
```rs
fn impl_trait(generics: ast::GenericParamList, trait_: Option<ast::Path>, target: ast::Type, assoc_item_list: ast::AssocItemList) -> ast::Impl
```
and it should only do minimal work to create the `ast::Impl` node, so the caller will have to do a bit more work in creating the input nodes via `make` as well then.

Reply:
I don't think that signature is right, because the generics should be optional?